### ISSUE TITLE
External memory support

### ISSUE THREAD

**Issue by akrvx** [GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator#324]
Hello. This has been raised before, but are there any plans on supporting extensions like `VK_KHR_external_memory_win32` and `VK_KHR_external_memory_fd`? These are particularly useful for applications that communicate data with, for example, CUDA.

**Comment by adam-sawicki-a** [comment]
Hello. How would you like this support to look like? You can use VMA with external memory currently with help of multiple features, like:

- Custom pools and `VmaPoolCreateInfo::pMemoryAllocationNext` which allows you to attach a custom structure to every `VkDeviceMemory` allocated that VMA will make in this pool.
- Use `VmaAllocatorCreateInfo::pTypeExternalMemoryHandleTypes` to make VMA automatically adding `VkExportMemoryAllocateInfoKHR` with specific handle type flags for every allocation made from a specific memory type.
- If the memory shared this way requires special alignment, use `vmaCreateBufferWithAlignment` or a custom pool with specific `VmaPoolCreateInfo::minAllocationAlignment`.

**Comment by akrvx** [comment]
Hi there, thanks for the reply. I think that setting `VmaAllocatorCreateInfo::pTypeExternalMemoryHandleTypes` is a nice and simple solution - I shall try that. I may come back adding a few comments here if that's ok.

**Comment by akrvx** [comment]
Yep, I had missed that bit before.
Everything is working fine this way. I create a pool per image, though, for each image that needs to be backed by external memory. I guess this is all right.
Thanks for the help.

**Comment by adam-sawicki-a** [comment]
This is all right, but do you really need to create a separate pool for each image? Does each one need different parameters in this `pNext` chain structures? If not, please remember that you can also create dedicated allocations in custom pools, so if you need these images to each have its own device memory block, maybe you can still use a single pool and allocate them with `VMA_ALLOCATION_CREATE_DEDICATED_MEMORY_BIT `?

**Comment by mojocorp** [comment]
It's exactly the same exact problem I'm currently trying to solve. The security attributes and one pool per image :)
It would be really great to have the security attribute sample in the doc.

**Comment by sawickiap** [comment]
Please note that the latest VMA code on the master branch has new functions available:

[`vmaCreateDedicatedImage`](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/group__group__alloc.html#ga07c66ffa000906a599d471047a7c0324)
[`vmaCraeteDedicatedBuffer`](https://gpuopen-librariesandsdks.github.io/VulkanMemoryAllocator/html/group__group__alloc.html#gab313b89299877ca21c196027bed9e874)

That offer extra parameter `pMemoryAllocateNext` which allows attaching a custom `pNext` chain for the `VkMemoryAllocateInfo` structure. Would that help in your case?

Regarding adding a sample about security attributes to the documentation, I am not happy to do that. VMA already supports many features for external memory import/export, far more than I originally planned, considering I treat all this as a special case that should rather be implemented using raw Vulkan functions directly, while VMA is intended to help with typical cases where automatic memory type selection and sub-allocation from larger memory blocks is useful.

**Comment by mojocorp** [comment]
Yes the dedicated allocator helps, thanks you so much.
I can now export/import a win32 opaque handle in the same process without issue. 
(I'm still not able to import it from another process yet, vmaCreateDedicatedImage fails, but that's another story (wink at @christosa-rfpro) )
